server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java

```diff
@@ -59,11 +58,11 @@ public DocumentMapper(RootObjectMapper.Builder rootBuilder, MapperService mapper
 
         final Collection<String> deleteTombstoneMetadataFields = Arrays.asList(VersionFieldMapper.NAME, IdFieldMapper.NAME,
             SeqNoFieldMapper.NAME, SeqNoFieldMapper.PRIMARY_TERM_NAME, SeqNoFieldMapper.TOMBSTONE_NAME);
-        this.deleteTombstoneMetadataFieldMappers = Stream.of(mapping.metadataMappers)
+        this.deleteTombstoneMetadataFieldMappers = Stream.of(mapping.getSortedMetadataMappers())
             .filter(field -> deleteTombstoneMetadataFields.contains(field.name())).toArray(MetadataFieldMapper[]::new);
         final Collection<String> noopTombstoneMetadataFields = Arrays.asList(
             VersionFieldMapper.NAME, SeqNoFieldMapper.NAME, SeqNoFieldMapper.PRIMARY_TERM_NAME, SeqNoFieldMapper.TOMBSTONE_NAME);
-        this.noopTombstoneMetadataFieldMappers = Stream.of(mapping.metadataMappers)
+        this.noopTombstoneMetadataFieldMappers = Stream.of(mapping.getSortedMetadataMappers())
             .filter(field -> noopTombstoneMetadataFields.contains(field.name())).toArray(MetadataFieldMapper[]::new);
```

Comment:
I'm pretty sure all this can just be replaced with a static set of an IdFieldMapper, SeqNoFieldMapper and VersionFieldMapper, but that can be done in a followup.

Reply:
yes this is all a bit confusing, it seems very delicate too as in some places we need them sorted, in others we need the map by class, in others we retrieve them by name. I also wonder if we can further clean up DocumentMapper around retrieving metadata mappers.